stm32f412 and stm32f413 have a 48MHz clock for SDMMC or USB

FRASTM opened the pull request:
Add the clock configuration for the 48MHz for the stm32f412 or stm32f413 devices
Both have a clock 48MHz which source is PLL Q or PLL I2S Q 
This 48MHz clock is feeding the USB or RNG or SDMCC peripherals

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/78816

FRASTM committed:
drivers: clock control: stm32F412 has PLL48MHz

Add the configuration of the PLL Q divider of main PLL
and I2S_Q of the PLLI2S toset the PLL48MHz clock which feeds
 the USB, SDMMC, RNG through the RCC_DCKCFGR2 register.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: pll i2s for the stm32f412 has a Q divider

There is a Q-divider factor [2..15] for the stm32f412 serie
which supplies the 48MHz clock.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: arm: stm32f412 device has a clock 48MHz multiplexer

Add a clk48Mhz node to the stm32f412 serie.
This clock is sourced by PLL_Q (default) or PLLI2S_Q
That 48MHz clock is used by the USB /SDMMC/RNG peripherals.
The sdmmc/SDIO clock is sourced by this CK48 (default)
or by the SYSCLOCK.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: clock: stm32 common device for sdmmc

Add a testcase for the stm32F412 or stm32F413
configuring the SDIO clock at 48MHz from the PLLI2S
Tested on the stm32f413h disco kit.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
from the stm32f412 [refman](https://www.st.com/resource/en/reference_manual/rm0402-stm32f412-advanced-armbased-32bit-mcus-stmicroelectronics.pdf):

![stm32f412_clock](https://github.com/user-attachments/assets/b35ac750-9fb2-4212-9f28-101b3ca4ef9d)

tagunil commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -215,6 +215,13 @@ int enabled_clock(uint32_t src_clk)
+			r = -ENOTSUP;
+		}
+		break;
+#endif /* STM32_SRC_PLLI2S_R */
```
```suggestion
#endif /* STM32_SRC_PLLI2S_Q */
```

FRASTM replied:
done

FRASTM commented:
Adding a testcase for the 48MHz clock configuration

`west build -p auto -b stm32f412g_disco tests/drivers/clock_control/stm32_clock_configuration/stm32_common_devices/ -- -DDTC_OVERLAY_FILE="boards/f4_sdmmc48_pll.overlay"`

Tested on a **stm32f413h_disco**,
```
Running TESTSUITE stm32_common_devices_clocks
===================================================================
START - test_sdmmc_clk_config
SDMMC gating clock on
SDMMC domain clock configured
SDMMC sourced by PLLI2SQ at 48000000 Hz
SDMMC clock rate: 48000000 Hz
SDMMC gating clk off
 PASS - test_sdmmc_clk_config in 0.013 seconds
===================================================================
START - test_sysclk_freq
 PASS - test_sysclk_freq in 0.001 seconds
===================================================================
TESTSUITE stm32_common_devices_clocks succeeded

------ TESTSUITE SUMMARY START ------

SUITE PASS - 100.00% [stm32_common_devices_clocks]: pass = 2, fail = 0, skip = 0, total = 2 duration = 0.014 seconds
 - PASS - [stm32_common_devices_clocks.test_sdmmc_clk_config] duration = 0.013 seconds                                                                                            
 - PASS - [stm32_common_devices_clocks.test_sysclk_freq] duration = 0.001 seconds                                                                                                 
                                                                                                                                                                                  
------ TESTSUITE SUMMARY END ------                                                                                                                                               
                                                                                                                                                                                  
===================================================================                                                                                                               
PROJECT EXECUTION SUCCESSFUL 
```

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32f410_clock.h:
```diff
@@ -6,6 +6,9 @@
 #ifndef ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32F410_CLOCK_H_
 #define ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32F410_CLOCK_H_
 
+/* CLK48MHz sources */
+#define STM32_SRC_CK48	(STM32_SRC_PLLI2S_R + 1)
+
```
This should be added in stm32f4_clock.h. With this single define in a different file, there is a great risk that it would be overlooked if any other clock sources are added in the future (PLLSAI for example).
There is no risk to have that define even for SoC that don't have this clock (just like F410 that doesn't have a PLLI2S but has the associated define).

gautierg-st commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -114,7 +114,16 @@ void config_plli2s(void)
 				       plli2sm(STM32_PLLI2S_M_DIVISOR),
 				       STM32_PLLI2S_N_MULTIPLIER,
 				       plli2sr(STM32_PLLI2S_R_DIVISOR));
+#if STM32_PLLI2S_Q_ENABLED
+	/* There is a Q divider on the PLLI2S to configure the PLL48CK */
+	LL_RCC_PLLI2S_ConfigDomain_48M(get_pll_source(),
+				       plli2sm(STM32_PLLI2S_M_DIVISOR),
+				       STM32_PLLI2S_N_MULTIPLIER,
+				       plli2sq(STM32_PLLI2S_Q_DIVISOR));
+#endif /* STM32_PLLI2S_Q_ENABLED */
```
There is currently no configuration of main PLL Q channel. You should also add the LL_RCC_PLL_ConfigDomain_48M function in the config_pll_sysclock function.

FRASTM replied:
adding a LL_RCC_PLL_ConfigDomain_48M function in the config_pll_sysclock() function

gautierg-st commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -428,6 +431,12 @@
 #endif
 
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk48), st_stm32_clock_mux, okay)
+/* Assuming the 48MHz clock is 48MHz */
+#define STM32_CK48_ENABLED	1
+#define STM32_CK48_FREQ		48000000
```
Although it is named 48MHz clock, it can technically be another value so it would be safer to compute its value.
This computed value could then be used to throw an assert if not equal to 48MHz.

FRASTM replied:
We can imagine a get_ck48_frequency() function to return the clock freq depending on the clock source (CK48MSEL )  of the PLL48 to be part of the stm32_clock_control_get_subsys_rate (get_rate). In this case I do not see any difference between the _dev_dt_clk_freq_ and the _dev_actual_clk_freq_   
--> actually there is no more _dev_dt_clk_freq_  (except if we add a clock-frequency property to the "st,stm32-clock-mux")

FRASTM committed:
drivers: clock control: stm32 function to get 48MHz freq

Add a function to compute the clock48 from the clock tree
of a stm32f412/f413 mcu. The value depends on its clock source
Requires to identify the PLL source HSE or HSI.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
with new commit 7ffa15a5dbc6a87b3cd763511a2be6a66441afb2 to get the 48MHz freq

gautierg-st commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -50,6 +50,36 @@ uint32_t get_pllsrc_frequency(void)
+	if (LL_RCC_GetCK48MClockSource(LL_RCC_CK48M_CLKSOURCE) ==
+			LL_RCC_CK48M_CLKSOURCE_PLL) {
+		/* Get the PLL Q freq. No HAL macro for that */
+		return __LL_RCC_CALC_PLLCLK_48M_FREQ(HSE_VALUE,
```
This supposes that HSE is used, but HSI could also be used as PLL source clock.
Function RCC_PLL_GetFreqDomain_48M could be used instead.

FRASTM replied:
could it be LL_RCC_PLL_GetMainSource

gautierg-st commented on drivers/clock_control/clock_stm32f2_f4_f7.c:
```diff
@@ -50,6 +50,36 @@ uint32_t get_pllsrc_frequency(void)
+	} else if (LL_RCC_GetCK48MClockSource(LL_RCC_CK48M_CLKSOURCE) ==
+			LL_RCC_CK48M_CLKSOURCE_PLLI2S) {
+		/* Get the I2S PLL Q freq. No HAL macro for that */
+		return __LL_RCC_CALC_PLLI2S_48M_FREQ(HSE_VALUE,
```
This supposes that HSE is used, but HSI could also be used as PLL source clock.
Function RCC_PLLI2S_GetFreqDomain_48M could be used instead.

FRASTM replied:
could it be LL_RCC_PLLI2S_GetMainSource ?

FRASTM commented:
rebase

erwango commented on dts/arm/st/f4/stm32f412.dtsi:
```diff
@@ -201,8 +207,7 @@
 		sdmmc1: sdmmc@40012c00 {
-			clocks = <&rcc STM32_CLOCK(APB2, 11U)>,
-				 <&rcc STM32_SRC_SYSCLK SDIO_SEL(1)>;
+			clocks = <&rcc STM32_CLOCK(APB2, 11U)>;
```
```
			clocks = <&rcc STM32_CLOCK(APB2, 11U)>,
				 <&rcc STM32_SRC_48M SDIO_SEL(0)>;
```

erwango commented on dts/arm/st/f4/stm32f412.dtsi:
```diff
@@ -16,6 +16,12 @@
+		clk48: clk48 {
+			#clock-cells = <0>;
+			compatible = "st,stm32-clock-mux";
+			status = "disabled";
```
clocks = <&rcc STM32_SRC_PLL_Q CK48M_SEL(0)>;

FRASTM commented:
adding the clock source default position for the CLK48MSEL and for the SDIOSEL

erwango commented on include/zephyr/drivers/clock_control/stm32_clock_control.h:
```diff
@@ -428,6 +431,12 @@
+#if DT_NODE_HAS_COMPAT_STATUS(DT_NODELABEL(clk48), st_stm32_clock_mux, okay)
+/* Assuming the 48MHz clock is 48MHz */
+#define STM32_CK48_ENABLED	1
+#define STM32_CK48_FREQ		48000000
```
Actually not used. Let remove

FRASTM replied:
removed

erwango reviewed (changes requested):
Otherwise LGTM

FRASTM commented on tests/drivers/clock_control/stm32_clock_configuration/stm32_common_devices/testcase.yaml:
```diff
@@ -34,3 +34,6 @@ tests:
     platform_allow: stm32f3_disco
+  drivers.clock.stm32_clock_configuration.common_device.f4.sdmmc_48:
+    extra_args: DTC_OVERLAY_FILE="boards/f4_sdmmc48_pll.overlay"
+    platform_allow: stm32f412g_disco nucleo_f412zg
```
changing to 2 lines
